src/OpenApi/src/Transformers/Implementations/OpenApiSchemaReferenceTransformer.cs

```diff
@@ -0,0 +1,137 @@
+{
+    public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
+    {
+        var schemaStore = context.ApplicationServices.GetRequiredKeyedService<OpenApiSchemaStore>(context.DocumentName);
+        var schemasByReference = schemaStore.SchemasByReference;
+
+        foreach (var pathItem in document.Paths.Values)
+        {
+            for (var i = 0; i < OpenApiConstants.OperationTypes.Length; i++)
+            {
+                var operationType = OpenApiConstants.OperationTypes[i];
+                if (pathItem.Operations.TryGetValue(operationType, out var operation))
+                {
+                    if (operation.Parameters is not null)
+                    {
+                        foreach (var parameter in operation.Parameters)
+                        {
+                            parameter.Schema = ResolveReferenceForSchema(parameter.Schema, schemasByReference);
+                        }
+                    }
+
+                    if (operation.RequestBody is not null)
+                    {
+                        foreach (var content in operation.RequestBody.Content)
+                        {
+                            content.Value.Schema = ResolveReferenceForSchema(content.Value.Schema, schemasByReference);
+                        }
+                    }
+
+                    if (operation.Responses is not null)
+                    {
+                        foreach (var response in operation.Responses.Values)
+                        {
+                            if (response.Content is not null)
+                            {
+                                foreach (var content in response.Content)
+                                {
+                                    content.Value.Schema = ResolveReferenceForSchema(content.Value.Schema, schemasByReference);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        document.Components ??= new OpenApiComponents();
+        document.Components.Schemas ??= new Dictionary<string, OpenApiSchema>();
+
+        foreach (var (schema, referenceId) in schemasByReference.OrderBy(kvp => kvp.Value))
+        {
+            if (referenceId is not null)
+            {
+                document.Components.Schemas[referenceId] = ResolveReferenceForSchema(schema, schemasByReference, skipResolution: true);
+            }
+        }
+
+        return Task.CompletedTask;
+    }
+
+    internal static OpenApiSchema? ResolveReferenceForSchema(OpenApiSchema? schema, ConcurrentDictionary<OpenApiSchema, string?> schemasByReference, bool skipResolution = false)
+    {
+        if (schema is null)
+        {
+            return schema;
+        }
+
+        if (!skipResolution && schemasByReference.TryGetValue(schema, out var referenceId) && referenceId is not null)
```

Comment:
This is probably answered elsewhere, but why would there be a null `referenceId` in the map?

Reply:
This is explained in the code comment on [PopulateSchemaIntoReferenceCache](https://github.com/dotnet/aspnetcore/blob/0a8f9c556e9b0bf3ff2a21f25df9ec472e0f8df8/src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs#L54) but that gist is that we only set a reference ID for a schema the second time we encounter it in the cache. This way, we effectively only generate refs for non-unique schemas.

Reply:
I think you're saying that null is a placeholder until the schema is seen a second time?  If so, I didn't get that from the comment you linked.

Reply:
I updated the comment here to be a little clearer...